### PR TITLE
Add min in wrap/unwrap/supply

### LINKED ISSUE #843: [Bulker UX] wrap and unwrap amount with a withdraw / repay
## Situation
Morpho is a lending protocol accruing interests at each block. So the balance of a collateral / debt is hard to know in advance, even impossible since you have to know the timestamp of the block including the tx.

The problem comes when you want to do actions after a max withdraw/repay with the retrieved tokens.

For example, Alice wants to withdraw all her wstEth collateral and unwrap everything. To withdraw, we are just going to call withdraw(wstEth, address(bulker), type(u256).max). The bulker has now all the collateral. And what is the amount to unwrap?

Another example, Bob wants to repay all his weth debt with native eth. So we send eth to the contract, the bulker wrap all the balance needed to repay with a buffer (to repay the interests accrued since the last amount was computed). so Morpho is going to take only the needed balance. This behavior leads to a small amount of weth remaining in the bulker. We can call skim(wEth), but a repay with a native eth leads to a transfer of wEth to Bob, which is not expected at all. With the current implémentation, we cannot do anything else.

Another example, we know that wrapping and unwrapping stEth can lead to 1 wei less. So the supply from a stEth can only be done by letting potentially 1 wei of wstEth in the bulker, which is not really clean as an onchain implementation

### Do nothing
We let some dust in the bulker in case of a max withdraw etc. This can lead to unexpected reverts if balances computed are not fully accrued.

### Add a max limiter
Limit the amounts with the balance of the bulker.

Let's take the example of the `_unwrapStEth` function:

```solidity
    function _unwrapStEth(bytes calldata data) internal {
        (uint256 amount, address receiver) = abi.decode(data, (uint256, address));

        amount = Math.min(amount, ERC20(_WST_ETH).balanceOf(address(this)));

        if (amount == 0) revert AmountIsZero();
        if (receiver == address(this)) revert TransferToSelf();


        uint256 unwrapped = IWSTETH(_WST_ETH).unwrap(amount);

        ERC20(_ST_ETH).safeTransfer(receiver, unwrapped);
    }
```

That also permits to homogenize the behavior with the one of Morpho where the withdraw//repay are maximized by the balance of the user

### PR DESCRIPTION
**Pull request by MerlinEgalite**
# Pull Request

## Issue(s) fixed

This pull request:

- fixes #843

### CODE REVIEW

**Review by Rubilmax — DISMISSED** [review]

File: test/integration/TestIntegrationBulkerGateway.sol [hunk 1]
```diff
@@ -160,10 +160,10 @@ contract TestIntegrationBulkerGateway is IntegrationTest {
         (actions[0], data[0]) = _getWrapStETHData(amount);
 
         bulker.execute(actions, data);
-        assertEq(
-            ERC20(stNative).balanceOf(address(bulker)), IWSTETH(stNative).getWstETHByStETH(amount), "bulker balance"
+        assertApproxEqAbs(
+            ERC20(stNative).balanceOf(address(bulker)), IWSTETH(stNative).getWstETHByStETH(amount), 1, "bulker balance"
         );
-        assertEq(ERC20(stETH).balanceOf(address(bulker)), 0, "bulker balance");
+        assertApproxEqAbs(ERC20(stETH).balanceOf(address(bulker)), 0, 1, "bulker balance");
```
Comment (julien-devatom) [inline test/integration/TestIntegrationBulkerGateway.sol:166]: Can you add a comment about why its an approximation at 1wei now?

**Review by julien-devatom — COMMENTED** [review]
IMO you should revert with AmountIsZero() if the min is zero, like that: 

```solidity
  function _unwrapStEth(bytes calldata data) internal {
        (uint256 amount, address receiver) = abi.decode(data, (uint256, address));

        amount = Math.min(amount, ERC20(_WST_ETH).balanceOf(address(this)));

        if (amount == 0) revert AmountIsZero();
        if (receiver == address(this)) revert TransferToSelf();


        uint256 unwrapped = IWSTETH(_WST_ETH).unwrap(amount);

        ERC20(_ST_ETH).safeTransfer(receiver, unwrapped);
    }
```

**Review by Tristan22400 — CHANGES_REQUESTED** [review]
I am wondering why you did not add the same thing to the repay function. 
https://github.com/morpho-org/morpho-aave-v3/blob/f3aebab4881242730994e7025cf6bd51b31c0c55/src/extensions/BulkerGateway.sol#L197

File: src/extensions/BulkerGateway.sol [hunk 2]
```diff
@@ -230,15 +231,15 @@ contract BulkerGateway is IBulkerGateway {
...
+        SafeTransferLib.safeTransferETH(receiver, Math.min(amount, address(this).balance));
     }
 
     /// @dev Wraps the given input of stETH to wstETH.
     function _wrapStEth(bytes calldata data) internal {
         (uint256 amount) = abi.decode(data, (uint256));
         if (amount == 0) revert AmountIsZero();
 
-        IWSTETH(_WST_ETH).wrap(amount);
+        IWSTETH(_WST_ETH).wrap(Math.min(amount, ERC20(_ST_ETH).balanceOf(address(this))));
```
Comment (Tristan22400) [inline src/extensions/BulkerGateway.sol:242]: maybe it makes more sense to check if the amount is zero after taking the minimum on the function. What do you think ?

**Comment by MerlinEgalite** [comment]
> I am wondering why you did not add the same thing to the repay function.
> 
> https://github.com/morpho-org/morpho-aave-v3/blob/f3aebab4881242730994e7025cf6bd51b31c0c55/src/extensions/BulkerGateway.sol#L197

Because it takes the max of (amount, debt) so I would expect pass an infinite amount instead of a specific amount. I can change this though, @julien-devatom wdyt?

File: src/extensions/BulkerGateway.sol (hunk 2 above)
Comment (makcandrov) [inline src/extensions/BulkerGateway.sol:242]: I think it's better to take the minimum before the null check (here and everywhere else too)

**Review by Tristan22400 — DISMISSED** [review]

**Review by pakim249CAL — APPROVED** [review]

**Review by julien-devatom — APPROVED** [review]
🚀 🔥 LGTM 🔥 🚀

**Review by Rubilmax — APPROVED** [review]

**Pull request merged**